simulator/infra/cache/cache_tag_array.cpp

```diff
@@ -1,177 +1,147 @@
-/*
+/**
  * cache_tag_array.cpp
  * Implementation of the cache tag array model.
- * MIPT-MIPS Assignment 5.
- * Ladin Oleg.
+ * @author Oleg Ladin, Denis Los
  */
 
-/* C generic modules */
-#include <cstdlib>
+#include <cassert>
 
-/* C++ generic modules */
-#include <iostream>
 
-/* MIPT-MIPS infra */
+#include <infra/cache/cache_tag_array.h>
 #include <infra/macro.h>
 
-/* MIPT-MIPS modules */
-#include "cache_tag_array.h"
 
-LRUInfo::LRUInfo( uint32 ways, uint32 sets) : lru( sets)
+
+void LRUCacheInfo::touch( std::size_t key)
 {
-    std::list< uint32> l;
-    for ( uint32 i = 0; i < ways; ++i)
-    {
-        l.push_front( i);
-    }
-    std::fill_n( lru.begin(), sets, l);
+    auto lru_it = lru_hash.find( key);
+
+    if ( lru_it != lru_hash.end())
+        lru_list.splice( lru_list.begin(), lru_list, lru_it->second);
 }
 
-/* On hit - mark (push front) way that contains the set */
-void LRUInfo::touch( uint32 set, uint32 way)
+
+
+std::size_t LRUCacheInfo::update()
 {
-    auto& list = lru[ set];
-    for ( auto it = list.begin(); it != list.end(); ++it)
+    std::size_t return_value = number_of_elements;
+
+    if ( number_of_elements == CAPACITY)
```

Comment:
I think you may initialize the hash table in ctor and remove `number_of_elements`